cmd/tiller-proxy/internal/handler/handler.go

```diff
@@ -126,7 +126,11 @@ func getChart(req *http.Request, cu chartUtils.Resolver) (*chartUtils.Details, *
 	if err != nil {
 		return nil, nil, err
 	}
-	ch, err := cu.GetChart(chartDetails)
+	netClient, err := cu.InitNetClient(chartDetails.Auth.CustomCA)
```

Comment:
Does this mean that for every single request made to get the chart we are going to initialize the client and hence read a secret for Kubernetes? If that's the case it seems an unacceptable overload

Reply:
We might want to have the ca's preloaded or add an additional cache layer so it is fetched only once?

Reply:
Actually just noticed that we are already reading the secret in GetChart which means that the secret wiill be fetched twice in the case of both header and CA are set no?

Would it be possible to preload this or cache it?

Reply:
We shouldn't assume in tiller-proxy that the auth header and custom CA Secrets are the same, given they both each have their own secretKeyRefs and if we change the way we handle the Secrets in the future we don't need to worry about this.

I think a cache layer (that caches by key to avoid above issue) that fetches if it isn't loaded might be a good idea - but what happens if the Secret gets updated? When there's  a cache hit, we could return the cached value and then spin off a goroutine to get the updated Secret, but that would still cause it to fail the first time it tries to install a chart with an updated auth header/custom CA.

Reply:
> We shouldn't assume in tiller-proxy that the auth header and custom CA Secrets are the same, given they both each have their own secretKeyRefs and if we change the way we handle the Secrets in the future we don't need to worry about this.

That's fine as long as there is some caching involve.

> I think a cache layer (that caches by key to avoid above issue) that fetches if it isn't loaded might be a good idea - but what happens if the Secret gets updated?

Those are difficult problems indeed. What you mentioned about checking the secret maybe via the resourceID in a go-rountine could be enough in most cases. Alternatively, our tiller-proxy `details` could contain the sha or resourceID of the provided CustomCA, so we will compare it with the one cached. This will allow us to pre-load the data on boot but also re-fetch it if stale,

What I am worried here is about adding a synchronous operation to an already costly op (deploy or upgrade). Do we know how long it takes to pull the secret?

I am going to remove my "required changes" for this PR because it seems that we are just building on top of an already existing pattern and we are affected only when auth is set, but I'd love to see a follow up improvement here.

Reply:
https://github.com/kubeapps/kubeapps/issues/895

Reply:
> Alternatively, our tiller-proxy details could contain the sha or resourceID of the provided CustomCA

Interesting idea, that would require the frontend to pull down the Secret to compute the checksum though, right? In that case, wouldn't it be okay to simply the pass the header and custom CA directly in the details? Assuming we talk to tiller-proxy over TLS, it's no less secure than us already passing the Kubernetes Bearer token in every request.

Reply:
I am not completely familiar with how things work now but yes, somehow the sha would need to be generated and potentially stored in the appRepository CRD as annotation or field.

We could make the appRepository controller be in charge of populating that value too?